vowpalwabbit/parser/flatbuffer/parse_flat_example.cc

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include <fstream>
+#include <iostream>
+#include <cfloat>
+
+#include "../../global_data.h"
+#include "../../example.h"
+#include "../../constant.h"
+#include "../../cb.h"
+#include "../../action_score.h"
+#include "../../best_constant.h"
+#include "parse_flat_example.h"
+#include "generated/example_generated.h"
+
+namespace VW {
+namespace parsers {
+namespace flatbuffer {
+
+void read_flatbuffer(vw* all, char* line, size_t len, v_array<example*>& examples)
+{
+  // Do something
+}
+
+int flatbuffer_to_examples(vw* all, v_array<example*>& examples)
+{
+  return (int) all->flat_converter->parse_examples(all, examples); // Get rid of this
+}
+
+parser::parser(std::string filename) : 
+_filename(filename),
+_flatbuffer_pointer(nullptr),
+_example_index(0),
+_c_hash(0)
+{}
+
+parser::parser(uint8_t *buffer_pointer) : 
+_filename("empty"),
+_flatbuffer_pointer(buffer_pointer),
+_example_index(0),
+_c_hash(0)
+{
+  _data = VW::parsers::flatbuffer::GetExampleCollection(_flatbuffer_pointer); //Shift as a separate call
+}
+
+void parser::init()
+{
+  std::ifstream infile;
+  infile.open(_filename, std::ios::binary | std::ios::in);
+  if (!infile.good()) THROW_EX(VW::vw_argument_invalid_value_exception, "Flatbuffer does not exist");
+
+  infile.seekg(0,std::ios::end);
+  int length = infile.tellg();
+  infile.seekg(0,std::ios::beg);
+  std::unique_ptr<char> buffer_pointer(new char[length]);
+  infile.read(buffer_pointer.get(), length);
+  _flatbuffer_pointer = reinterpret_cast<u_int8_t*>(buffer_pointer.get());
+  infile.close();
+
+  _data = VW::parsers::flatbuffer::GetExampleCollection(_flatbuffer_pointer);
+}
+
+const VW::parsers::flatbuffer::ExampleCollection* parser::data() {return _data;}
+
+bool parser::parse_examples(vw* all, v_array<example*>& examples)
+{
+  if (!flatbuffers::IsFieldPresent(_data, ExampleCollection::VT_EXAMPLES))
+  {
+    THROW("No examples to parse");
+  }
+  if (_example_index == _data->examples()->size())
+    return false;
+  parse_example(all, examples[0], _data->examples()->Get(_example_index));
+  _example_index++;
+  return true;
+}
+
+void parser::parse_example(vw* all, example* ae, const Example* eg)
+{
+  all->p->lp.default_label(&ae->l);
+  parse_flat_label(all->sd, ae, eg);
+
+  if (flatbuffers::IsFieldPresent(eg, Example::VT_TAG)){
+    ae->tag = v_init<char>();
+    VW::string_view tag(eg->tag()->c_str());
+    push_many(ae->tag, tag.begin(), tag.size());
+  }
+
+  for (int i=0; i < eg->namespaces()->size(); i++){
+    parse_namespaces(all, ae, eg->namespaces()->Get(i));
+  }
+}
+
+void parser::parse_namespaces(vw* all, example* ae, const Namespace* ns)
+{
```

Comment:
`ae->indices` is a `std::array` of `namespace_index`'s. You could declare `temp_index` as a `namespace_index` here (and there will be no need for the `uint8_t` cast of `ns->name()`). Also `ae->indices.push_back(tmp_index)` could be done once outside of the if/else clause.